### PR TITLE
Enhancements on new regridding code

### PR DESCRIPTION
**Pull request by zfan001**
Enhancements on new regridding code
(1) Considering cases with target grid domain bigger than source grid domain. It happens when ACCESS city model is included in BOM's solution. Another occasion is  that some Albers equal-area projection  Grid points are outside MSAS/NAS input domain. 
(2) If input cube's coordinates descend, reverse it to ascending before regridding instead of stopping with an error.
(3) Ensure the same array type (masked or not) between input cube's data  and  output cube's data.
(4) Several tests and verifications jupyter notebooks will be merged into improver_aux.
(5) These enhancements are applicable for nearest-neighbour (nearest-2), bilinear (bilinear-2), land-sea-aware nearest-neighbour(nearest-with-mask-2), land-sea-aware bilinear(bilinear-with-mask-2) options.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: improver/regrid/grid.py
Comment (dmentipl) [inline improver/regrid/grid.py:464]: Can you add a blank line before `Returns:`.

File: improver/regrid/landsea2.py
Comment (dmentipl) [inline improver/regrid/landsea2.py:161]:
```suggestion
        # exclude out-of-input-domain target point here
```

File: improver/regrid/grid.py
Comment (dmentipl) [inline improver/regrid/grid.py:469]: Just a comment: a somewhat more concise way of expressing this line is

```python
    output_shape = (total_out_point_num, *regrid_result.shape[1:])
```

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (dmentipl) [inline improver_tests/regrid/test_RegridWithLandSeaMask.py:346]:
```suggestion
    # add a circle of grid points so that output domain is much bigger than input domain
```

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (dmentipl) [inline improver_tests/regrid/test_RegridWithLandSeaMask.py:379]:
```suggestion
    # add a circle of grid points so that output domain is much bigger than input domain
```

**Review by dmentipl — COMMENTED** [review]
Thanks for this PR @zfan001. I've made a few minor comments. I'm not an expert on re-gridding. I expect other reviewers are better equipped to dig into the implementation more so that I have.

File: improver/regrid/bilinear.py
Comment (zfan001) [reply improver/regrid/bilinear.py:138]: Removed.  Thanks

File: improver/regrid/grid.py
Comment (zfan001) [reply improver/regrid/grid.py:98]: added

File: improver/regrid/grid.py
Comment (zfan001) [reply improver/regrid/grid.py:420]: added

File: improver/regrid/grid.py
Comment (zfan001) [reply improver/regrid/grid.py:417]: done

File: improver/regrid/grid.py
Comment (zfan001) [reply improver/regrid/grid.py:464]: done

File: improver/regrid/landsea2.py
Comment (zfan001) [reply improver/regrid/landsea2.py:161]: done

File: improver/regrid/grid.py
Comment (zfan001) [reply improver/regrid/grid.py:469]: thanks

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (zfan001) [reply improver_tests/regrid/test_RegridWithLandSeaMask.py:379]: done

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (zfan001) [reply improver_tests/regrid/test_RegridWithLandSeaMask.py:346]: done

File: improver/regrid/bilinear.py
Comment (zfan001) [reply improver/regrid/bilinear.py:55]: Thanks. Changes in this function is for the data array type consistency between input cube and output cube.

File: improver/regrid/bilinear.py
Comment (zfan001) [reply improver/regrid/bilinear.py:73]: Yes, using np.nan filled unmasked array for calculation can make calculation several times faster with no impact on results.

File: improver/regrid/bilinear.py
Comment (zfan001) [reply improver/regrid/bilinear.py:196]: This fixing was done for a bug for a Albers projection case, in which point_lat_lon_max_index is equal to 0.

File: improver/regrid/grid.py
Comment (zfan001) [reply improver/regrid/grid.py:58]: This fixing is done for EC dataset input, in which latitude coordinates are descending. It is only for input source cube and input mask cube.

File: improver/regrid/grid.py
Comment (zfan001) [reply improver/regrid/grid.py:478]: Good point. Done as suggested.

File: improver_tests/regrid/test_grid.py
Comment (zfan001) [reply improver_tests/regrid/test_grid.py:153]: There are two layers of protections for this case: 
(1) in "ensure_ascending_coord", we have :
if cube.coord(axis=ax).points[0] > cube.coord(axis=ax).points[-1]:
(2) sort_coord_in_cube only produce ascending coordinate by default.
I reckon that it is probably good enough without adding more tests.

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (zfan001) [reply improver_tests/regrid/test_RegridWithLandSeaMask.py:451]: Good catch, Thanks. I made changes.

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (zfan001) [reply improver_tests/regrid/test_RegridWithLandSeaMask.py:424]: It is nice to use parametrization for reducing duplicated code. Thanks!
"-1" on lines 386 and 394 is unexpected, and I am debugging to figure out why it happened.

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (zfan001) [reply improver_tests/regrid/test_RegridWithLandSeaMask.py:357]: good point. Already done in your function.

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (zfan001) [reply improver_tests/regrid/test_RegridWithLandSeaMask.py:359]: Already done in your function.

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (zfan001) [reply improver_tests/regrid/test_RegridWithLandSeaMask.py:424]: I debugged it, and fixed the problem. The problem is in the original cube setting. I removed "-1".

File: improver_tests/regrid/test_grid.py
Comment (zfan001) [reply improver_tests/regrid/test_grid.py:154]: added as suggested.

**Review by tjtg — COMMENTED** [review]
All comments from my review have been resolved and I think this looks good now.

We should wait for an additional review from someone from Met Office before approving/merging.

File: improver/regrid/landsea2.py
Comment (MoseleyS) [inline improver/regrid/landsea2.py:117]: Please update the doc-string to describe the expected behaviour when the output domain is not a subset of the input domain. Please also correct the error in the existing doc-string (`cube_our` does not exist!)

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (MoseleyS) [inline improver_tests/regrid/test_RegridWithLandSeaMask.py:373]:
```suggestion
    regridder = RegridLandSea(
        regrid_mode=regrid_mode, landmask=cube_in_mask, landmask_vicinity=250000000,
    )
    regrid_out = regridder(cube_in, cube_out_mask)
    regrid_out_pad = regridder(cube_in, cube_out_mask_pad)
```

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (MoseleyS) [inline improver_tests/regrid/test_RegridWithLandSeaMask.py:385]:
```suggestion
        assert regrid_out_pad.data.is_masked()
```

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (MoseleyS) [inline improver_tests/regrid/test_RegridWithLandSeaMask.py:392]:
```suggestion
        assert not regrid_out_pad.data.is_masked()
```

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (MoseleyS) [inline improver_tests/regrid/test_RegridWithLandSeaMask.py:382]: Is this true? The code doesn't appear to handle an extra column of numbers.

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (MoseleyS) [inline improver_tests/regrid/test_RegridWithLandSeaMask.py:391]: I expected the plugin to return a masked array here too, to differentiate between points that had data and points that did not. Using `np.nan` is likely to cause issues in data smoothing modules as the nans tend to proliferate.

**Review by MoseleyS — CHANGES_REQUESTED** [review]
Thanks @zfan001. Being able to handle cases where the target grid is larger than the source grid is useful. I've suggested some changes for you to think about.

File: improver/regrid/bilinear.py
Comment (zfan001) [reply improver/regrid/bilinear.py:71]: done. Thanks

File: improver/regrid/bilinear.py
Comment (zfan001) [reply improver/regrid/bilinear.py:196]: done. thanks

File: improver/regrid/landsea2.py
Comment (zfan001) [reply improver/regrid/landsea2.py:125]: Thanks. I feel that the case hardly happens, and ensure_ascending_coord function will skip the function sort_coord_in_cube directly for vast majority of cases.

File: improver/regrid/grid.py
Comment (zfan001) [reply improver/regrid/grid.py:424]: Thanks for suggestion.  I prefer to keep unchanged to avoid several changes elsewhere.

File: improver/regrid/landsea2.py
Comment (zfan001) [reply improver/regrid/landsea2.py:117]: done. thanks

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (zfan001) [reply improver_tests/regrid/test_RegridWithLandSeaMask.py:373]: done. thanks

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (zfan001) [reply improver_tests/regrid/test_RegridWithLandSeaMask.py:385]: done. thanks

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (zfan001) [reply improver_tests/regrid/test_RegridWithLandSeaMask.py:392]: done. thanks

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (zfan001) [reply improver_tests/regrid/test_RegridWithLandSeaMask.py:382]: old invalid comments. deleted. Thanks

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (zfan001) [reply improver_tests/regrid/test_RegridWithLandSeaMask.py:391]: just keep output/input consistent here. I reckon that the normal input should be masked.

**Comment by zfan001** [comment]
Hi @MoseleyS. Sorry for the delay.  I am on leave at the moment. Pushed now.
Would you please review four new notebooks in improver_aux (PR 285)  so that it can be merged ? 
Thanks, Julian

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (MoseleyS) [inline improver_tests/regrid/test_RegridWithLandSeaMask.py:372]: Please add a test to check that the output data are of the same dtype as the input data (both should be `float32`)

**Review by MoseleyS — CHANGES_REQUESTED** [review]
Hi @zfan001 

We have tried to use the principle of only supporting those cases we actually have, and of only outputting just enough information to maintain scientific integrity.
On this basis, I think supporting NaNs in the input data should be avoided. I also think that the right output from regridding this case is a masked array, with unfillable values set as masked. We would not expect the input data to be masked, there should be data at all points for most parameters.

Please can you modify the code so that regardless of whether the input cube contains masked or unmasked data, the output will contain masked data only if there are masked points present.
I think that my earlier suggestion around handling masks instead of index arrays would make this easier.

I also think that if we don't expect any of our inputs to contain NaN values, then we don't need to support them. As it is much more useful and consistent to only have Masks supported, I think we should only support Masks in this code too.
I would then suggest that if we DO find any data with NaNs instead of masks, we use our `standardise` CLI to convert them into masks. This functionality isn't currently there, because we haven't needed it yet, but that's where it should be.

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (zfan001) [reply improver_tests/regrid/test_RegridWithLandSeaMask.py:372]: @MoseleyS I have come back from leave. Thank you very much for your detailed comments. 
Some groups in BOM are interested in using the land-sea-mask-aware regridding, and I would like to recommend them to install IMPROVER and then use regridding function. I thought that to make the input and output array type consistent would be good. 
I will do what you suggested. Thanks,
Julian

**Comment by zfan001** [comment]
@MoseleyS I have come back from leave. Thank you very much for your detailed comments. 
Some groups in BOM are interested in using the land-sea-mask-aware regridding, and I would like to recommend them to install IMPROVER and then use regridding function. I thought that to make the input and output array type consistent would be good. 
I will do what you suggested. Thanks,  Julian

**Comment by zfan001** [comment]
@MoseleyS  I looked into the code.  I can use just one line to change from array with np.nan to masked array. "output = np.ma.masked_invalid(output)" . However, I notice that the change is probably needed for other Iris regridding options as well.
The current way in our code is consistent with what Iris regridding is doing. If a small input domain is regridded into a large output domain using Iris regridding (IMPROVER regridding's nearest & bilinear options),  unmasked input will generate a unmasked output with np.nan inside, and masked input will generate a masked output.  In some verification tests, I used it for verification. It may be better to leave it for now and fix it in another PR later ?? 

Regarding the data type (float32), there are some problems with test examples. I fixed it. There are some result changes due to rounding error. Another change is about equal-distance issues for nearest neighbour option (the selected source point may be different due to rounding error).

Kind regards, Julian

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (MoseleyS) [inline improver_tests/regrid/test_RegridWithLandSeaMask.py:415]: I've just spotted that my suggested name `regridder` has already been used in this function. Please rename either the earlier string object or this RegridLandSea object so that they are clearly separate.

**Review by MoseleyS — CHANGES_REQUESTED** [review]
Hi @zfan001. I only have a couple of small suggestions that will not take long to deal with, then I will approve this PR.

As an aside, the `landsea2.RegridWithLandSeaMask` plugin needs refactoring in the future. The `process` method contains everything, making it difficult to follow (a larger number of small methods would be better). Also, the `__init__` method does not check that the plugin has been given usable options (what happens if regrid_mode is invalid, or vicinity_radius is negative?) and the global variables `BILINEAR2` and `NEAREST2` are never used. There is some protection as this plugin is currently only accessed from one place, but it would be better if it took care of itself to prevent accidental misuse.

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (zfan001) [reply improver_tests/regrid/test_RegridWithLandSeaMask.py]: Done. Thanks

File: improver_tests/regrid/test_RegridWithLandSeaMask.py
Comment (zfan001) [reply improver_tests/regrid/test_RegridWithLandSeaMask.py]: Thank your very much for editing. Done.

**Comment by zfan001** [comment]
@MoseleyS  Two changes have been made. Thanks. Also thank you for raising several points for future changes.   
Kind regards, Julian

**Review by MoseleyS — APPROVED** [review]

**Comment by zfan001** [comment]
@MoseleyS @tjtg @dmentipl Thanks for your review. Kind regards, Julian

**Pull request merged**
